ykrt/src/compile/mod.rs

```diff
@@ -38,15 +52,35 @@ struct SendSyncConstPtr<T>(*const T);
 unsafe impl<T> Send for SendSyncConstPtr<T> {}
 unsafe impl<T> Sync for SendSyncConstPtr<T> {}
 
-struct Guard {
-    failed: u32,
-    code: Option<SendSyncConstPtr<c_void>>,
+pub struct Guard {
+    failed: AtomicU32,
+    ct: Mutex<Option<Arc<CompiledTrace>>>,
+}
+
+impl Guard {
+    pub fn inc(&self) {
+        let n = self.failed.load(Ordering::Relaxed);
+        self.failed.store(n + 1, Ordering::Relaxed);
+    }
+
+    pub fn failcount(&self) -> u32 {
+        self.failed.load(Ordering::Relaxed)
+    }
+
+    pub fn setct(&self, ct: Arc<CompiledTrace>) {
+        let _ = self.ct.lock().insert(ct);
+    }
+
+    pub fn getct(&self) -> Option<Arc<CompiledTrace>> {
+        self.ct.lock().as_ref().cloned()
```

Comment:
I get nervous with `clone`/`cloned` because if the inner type changes this will do something, but probably not what we want. If possible, can we do something like `self.ct.lock().map(|x| Arc::clone(x))` to avoid this?

Reply:
Fixed in b6f02d72f